src/cargo/util/toml/targets.rs

```diff
@@ -139,6 +146,15 @@ fn clean_lib(
         (Some(path), _) => package_root.join(&path.0),
         (None, Some(path)) => path,
         (None, None) => {
+            let bail = || {
+                bail!(
+                    "can't find library `{}`, rename file to `src/lib.rs` or specify lib.path",
+                    lib.name()
+                )
+            };
+            if edition >= Edition::Edition2018 {
+                return bail();
+            }
             let legacy_path = package_root.join("src").join(format!("{}.rs", lib.name()));
             if legacy_path.exists() {
```

Comment:
Hm, let's add `&& edition == 2015` to this condition, instead of introducing the `bail` closure?